chain-storage/src/iterator.rs

```diff
@@ -0,0 +1,147 @@
+use sled::Tree;
+
+/// Iterator over blocks. Starts from n-th ancestor of the given block.
+pub struct StorageIterator {
+    state: IteratorState,
+    to: Value,
+    block_info: Tree,
+    blocks: Tree,
+}
+
+enum IteratorState {
+    Permanent {
+        iter: data_pile::SeqNoIter,
+        current_length: u32,
+        stop_at_length: u32,
+    },
+    Volatile {
+        ids: Vec<Value>,
+    },
+}
+
+impl StorageIterator {
+    /// Create a new iterator. `from` and `to` values MUST have a path between
+    /// them and this needs to be checked before the iterator is started.
+    pub(crate) fn new(
+        to: Value,
+        distance: u32,
+        permanent_store: PermanentStore,
+        block_info: Tree,
+        blocks: Tree,
+    ) -> Result<Self, Error> {
+        let to_info = if let Some(to_info_bin) = block_info.get(to.as_ref())? {
+            BlockInfo::deserialize(to_info_bin.as_ref(), to.as_ref().len(), to.clone())
+        } else {
+            permanent_store
+                .get_block_info(to.as_ref())?
+                .ok_or(Error::BlockNotFound)?
+        };
+        assert!(
+            to_info.chain_length() + 1 >= distance,
+            "expected distance {} <= chain length {}",
+            distance,
+            to_info.chain_length() + 1
+        );
+        let from_length = to_info.chain_length() + 1 - distance;
+
+        let state = if let Some(_) = permanent_store.get_block_by_chain_length(from_length) {
+            IteratorState::Permanent {
+                iter: permanent_store.iter(from_length)?,
+                current_length: from_length,
+                stop_at_length: to_info.chain_length(),
+            }
+        } else {
+            IteratorState::Volatile {
+                ids: gather_blocks_ids(to.clone(), &block_info, from_length)?,
+            }
+        };
+
+        Ok(Self {
+            state,
+            to,
+            block_info,
+            blocks,
+        })
+    }
+}
+
+impl Iterator for StorageIterator {
+    type Item = Value;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        match &mut self.state {
+            IteratorState::Permanent {
+                iter,
+                current_length,
+                stop_at_length,
+            } => {
+                if current_length == stop_at_length {
+                    return None;
+                }
+                match iter.next() {
+                    Some(item) => {
+                        *current_length += 1;
+                        Some(Value::permanent(item))
+                    }
+                    None => {
+                        self.state = IteratorState::Volatile {
+                            ids: gather_blocks_ids(
+                                self.to.clone(),
+                                &self.block_info,
+                                *current_length,
+                            )
+                            .ok()?,
```

Comment:
As I said previously, exposing the iterator API on fallible operations is fraught with problems.
Here, the iterator will stop on failure, and the consumer can only distinguish the failure case from end-of-iterator case by looking into the block metadata to check if the blocks received are what was requested. Then there will be code that will not do that and cause mysterious havoc when some rare error occurs.

I think it's important to avoid offering an error-ignorant iterator as the easiest available API, despite the inconvenience. I'd rather make this iterator return a `Result` and deal with the inconvenience, possibly assisted with utilities like [fallible-iterator](https://docs.rs/fallible-iterator/0.2.0/fallible_iterator/).

Reply:
Yes, sounds like a better approach. I thought you were against having `Item = Result<T, E>` so I didn't do that. Probably, just misunderstanding.